PWGHF: Add a flag to the different LcpKpi decay channels

DelloStritto opened the pull request.

vkucera commented on Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:
```diff
@@ -314,6 +314,8 @@ DECLARE_SOA_DYNAMIC_COLUMN(MaxNormalisedDeltaIP, maxNormalisedDeltaIP, [](float
 // - ±LcToPKPi: Λc± → p± K∓ π±
 DECLARE_SOA_COLUMN(FlagMCMatchRec, flagMCMatchRec, int8_t); // reconstruction level
 DECLARE_SOA_COLUMN(FlagMCMatchGen, flagMCMatchGen, int8_t); // generator level
+DECLARE_SOA_COLUMN(FlagMCDecayChanRec, flagMCDecayChanRec, int8_t); // reconstruction level
+DECLARE_SOA_COLUMN(FlagMCDecayChanGen, flagMCDecayChanGen, int8_t); // generator level
```
Please add comments describing the purpose of these columns.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreator3Prong.cxx:
```diff
@@ -154,11 +154,23 @@ struct HFCandidateCreator3ProngMC {
+    int8_t DecayChannel = 0;
+    int8_t size;
+    std::vector<int> arrDaughIndex;
+    std::vector<int> arrPDGDaugh;
+    std::vector<int> arrPDGResonant1 = {2212, 313};
+    std::vector<int> arrPDGResonant2 = {2224, 321};
+    std::vector<int> arrPDGResonant3 = {3124, 211};
```
Are vectors needed here? It seems, their lengths are constant and therefore could be arrays instead.

DelloStritto replied:
Good point. Thanks. Indeed is better to use the arrays.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreator3Prong.cxx:
```diff
@@ -154,11 +154,23 @@ struct HFCandidateCreator3ProngMC {
+      DecayChannel = 0;
+      size = 0;
+      arrDaughIndex = {};
+      arrPDGDaugh = {0, 0, 0};
```
Why does this need 3 elements if it is used only for daughters of the resonant channels?

DelloStritto replied:
Because it is filled also for the non resonant case.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreator3Prong.cxx:
```diff
@@ -170,18 +182,38 @@ struct HFCandidateCreator3ProngMC {
           result = sign * LcToPKPi;
+
+          //Printf("Flagging the different Λc± → p± K∓ π± decay channels");
+          RecoDecay::getDaughters(particlesMC, indexRecLc, &arrDaughIndex, array{+kProton, -kKPlus, +kPiPlus}, 1);
```
Since you ask for direct daughters, I don't see the point of providing the PDG codes of daughters, as the function will always return all the direct daughters from the first stage of recursion.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreator3Prong.cxx:
```diff
@@ -170,18 +182,38 @@ struct HFCandidateCreator3ProngMC {
+
+          //Printf("Flagging the different Λc± → p± K∓ π± decay channels");
+          RecoDecay::getDaughters(particlesMC, indexRecLc, &arrDaughIndex, array{+kProton, -kKPlus, +kPiPlus}, 1);
+          size = arrDaughIndex.size();
+          for (auto iProng = 0; iProng < size; ++iProng) {
+            auto daughI = particlesMC.iteratorAt(arrDaughIndex[iProng]);
+            arrPDGDaugh[iProng] = daughI.pdgCode();
```
Why is this part executed regardless of the number of direct daughters? Shouldn't everything below `getDaughters` be executed only for the resonant decays, i.e. `if (arrDaughIndex.size() == 2)`?

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreator3Prong.cxx:
```diff
@@ -170,18 +182,38 @@ struct HFCandidateCreator3ProngMC {
+            arrPDGDaugh[iProng] = daughI.pdgCode();
+          }
+          if (size == 2) {
+            if ((std::abs(arrPDGDaugh[0]) == arrPDGResonant1[0] && std::abs(arrPDGDaugh[1]) == arrPDGResonant1[1]))
```
If you assign directly `arrPDGDaugh[iProng] = std:abs(daughI.pdgCode());`, you don't have to repeat the `std:abs` later anymore.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreator3Prong.cxx:
```diff
@@ -154,11 +154,23 @@ struct HFCandidateCreator3ProngMC {
+    int8_t size;
+    std::vector<int> arrDaughIndex;
+    std::vector<int> arrPDGDaugh;
+    std::vector<int> arrPDGResonant1 = {2212, 313};
+    std::vector<int> arrPDGResonant2 = {2224, 321};
+    std::vector<int> arrPDGResonant3 = {3124, 211};
```
Add comments describing which particles these PDG codes correspond to.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreator3Prong.cxx:
```diff
@@ -170,18 +182,38 @@ struct HFCandidateCreator3ProngMC {
       if (result == N3ProngDecays) {
         //Printf("Checking Λc± → p± K∓ π±");
-        if (RecoDecay::getMatchedMCRec(particlesMC, std::move(arrayDaughters), 4122, array{+kProton, -kKPlus, +kPiPlus}, true, &sign) > -1) {
+        auto indexRecLc = RecoDecay::getMatchedMCRec(particlesMC, std::move(arrayDaughters), 4122, array{+kProton, -kKPlus, +kPiPlus}, true, &sign, -1);
```
Is it needed to keep the tree depth unrestricted?

DelloStritto replied:
It can be fixed to 2. I already checked that the result is the same.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreator3Prong.cxx:
```diff
@@ -192,12 +224,28 @@ struct HFCandidateCreator3ProngMC {
       if (result == N3ProngDecays) {
         //Printf("Checking Λc± → p± K∓ π±");
-        if (RecoDecay::isMatchedMCGen(particlesMC, particle, 4122, array{+kProton, -kKPlus, +kPiPlus}, true, &sign)) {
+        auto isMatchedGenLc = RecoDecay::isMatchedMCGen(particlesMC, particle, 4122, array{+kProton, -kKPlus, +kPiPlus}, true, &sign, -1);
+        if (isMatchedGenLc) {
           result = sign * LcToPKPi;
+
+          //Printf("Flagging the different Λc± → p± K∓ π± decay channels");
+          RecoDecay::getDaughters(particlesMC, particle.globalIndex(), &arrDaughIndex, array{+kProton, -kKPlus, +kPiPlus}, 1);
+          size = arrDaughIndex.size();
+          for (auto jProng = 0; jProng < size; ++jProng) {
+            auto daughJ = particlesMC.iteratorAt(arrDaughIndex[jProng]);
+            arrPDGDaugh[jProng] = daughJ.pdgCode();
+          }
+          if (size == 2) {
+            if ((std::abs(arrPDGDaugh[0]) == arrPDGResonant1[0] && std::abs(arrPDGDaugh[1]) == arrPDGResonant1[1]))
+              DecayChannel = 1;
```
Same comments as for the rec. part.

vkucera committed:
RecoDecay: getDaughters: Skip PDG check for the original particle. (#5303)

Do not label the original particle as final, when its species matches the expected species of its daughter.
This could happen when calling this function for a quark that radiates a gluon and becomes its own daughter.
In general, there is no reason to check the PDG code of the original particle against the codes of the expected daughters, so skipping it speeds up this function.
